### PR TITLE
Call readiness - UI Library Sample

### PR DESCRIPTION
**Pull request by JamesBurnside**
## What

Sample for creating a call-readiness experience using UI library.

Functionality:

- Part 1: Browser Compatibility Check
- Part 2: Request device permissions
- Part 3: Device Setup (Having the user chose their desired camera and microphone)

### CODE REVIEW

File: ui-library-call-readiness/README.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+[docs-ui-library]: https://azure.github.io/communication-ui-library/
+[React]: https://reactjs.org/
+
+## Sample Preview
+
+![Gif showing sample running end to end](./example-user-flow.gif)
+
+## Code organization
+
+TBD
```
Comment (JamesBurnside) [inline ui-library-call-readiness/README.md:29]: update TBDs in this readme

File: ui-library-call-readiness/src/App.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import { DeviceSetup } from './pages/DeviceSetup';
+import { TestComplete } from './pages/TestComplete';
+
+// Initializing and registering icons should only be done once per app.
+initializeIcons();
+registerIcons({ icons: DEFAULT_COMPONENT_ICONS });
+
+type TestingState = 'runningPreCallChecks' | 'deviceSetup' | 'finished';
+
+const USER_ID = 'user1'; // Replace with an Azure Communication Services User ID
```
Comment (JamesBurnside) [inline ui-library-call-readiness/src/App.tsx:18]: update comment to say for production apps they would retrieve a userId from their running service with a link to a server sample that can do this

File: ui-library-call-readiness/src/components/LocalPreview.tsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  CameraButton,
+  useTheme,
+  MicrophoneButton
+} from '@azure/communication-react';
+import { Stack, mergeStyles, Text, ITheme } from '@fluentui/react';
+import { VideoOff20Filled } from '@fluentui/react-icons';
+import { useEffect } from 'react';
+import { useCameras, useLocalPreview } from '../helpers/deviceSetupHooks';
+
+export const LocalPreview = (props: {
```
Comment (JamesBurnside) [inline ui-library-call-readiness/src/components/LocalPreview.tsx:14]: each component needs a doc comment

File: ui-library-call-readiness/src/helpers/deviceSetupHooks.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+      },
+      {
+        scalingMode: 'Crop'
+      }
+    );
+  }, [callClient, state.deviceManager.selectedCamera]);
+
+  const stopLocalPreview = useCallback(() => {
+    if (!localPreview) {
+      console.warn('no local preview ti dispose');
```
Comment (dmceachernmsft) [inline ui-library-call-readiness/src/helpers/deviceSetupHooks.ts:102]: nit: typo in message

File: ui-library-call-readiness/src/helpers/deviceSetupHooks.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
+import { AudioDeviceInfo, VideoDeviceInfo } from "@azure/communication-calling";
+import { CallClientState, StatefulDeviceManager, useCallClient, VideoStreamRendererViewState } from "@azure/communication-react";
+import { useCallback, useEffect, useRef, useState } from "react";
+
+/** A helper hook to get and update microphone device information */
+export const useMicrophones = (): {
```
Comment (dmceachernmsft) [inline ui-library-call-readiness/src/helpers/deviceSetupHooks.ts:6]: I really like how the hooks are seperated here. Makes it very easy to understand how they are to be used.

File: ui-library-call-readiness/src/helpers/deviceSetupHooks.ts (hunk 4 above)
Comment (JamesBurnside) [reply ui-library-call-readiness/src/helpers/deviceSetupHooks.ts]: updated 👍

**Review by dmceachernmsft — APPROVED** [review]

File: ui-library-call-readiness/src/components/EnvironmentChecksComponent.tsx [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  const callClient = useCallClient();
+  useEffect(() => {
+    const runEnvironmentChecks = async (): Promise<void> => {
+
+      // First we will get the environment information from the calling SDK.
+      const environmentInfo = await checkEnvironmentSupport(callClient);
+
+      if (!environmentInfo.isSupportedPlatform) {
+        setCurrentCheckState('operatingSystemUnsupported');
+        // if the platform or operating system is not supported we'll stop here and display a modal to the user.
```
Comment (edwardlee-msft) [inline ui-library-call-readiness/src/components/EnvironmentChecksComponent.tsx:35]: nit: Some comments start capitalized while some do not. 
ex. if vs If

**Review by edwardlee-msft — APPROVED** [review]

**Pull request merged**